pkg/pgmodel/metadata/metadata.go

```diff
@@ -0,0 +1,215 @@
+		if err != nil {
+			return nil, fmt.Errorf("fetching metadata: %w", err)
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var metricFamily, typ, unit, help string
+			if err := rows.Scan(&metricFamily, &typ, &unit, &help); err != nil {
+				return nil, fmt.Errorf("query result: %w", err)
+			}
+			data = append(data, model.Metadata{
+				MetricFamily: metricFamily,
+				Unit:         unit,
+				Type:         typ,
+				Help:         help,
+			})
+		}
+		var targetInfo map[string]map[Target]struct{}
+		targetInfo, err = getTargetInfo(conn, metrics, seriesIds)
+		if err != nil {
+			return nil, fmt.Errorf("get target info: %w", err)
+		}
+		for i := range data {
+			meta := data[i]
+			metricName := meta.MetricFamily
+			targetInfos := targetInfo[metricName]
+			for j := range targetInfos {
+				if targetLimit != 0 && len(metadataSlice) >= targetLimit {
+					break
+				}
+				metadataSlice = append(metadataSlice, TargetMetadataType{
+					Target:   j,
+					Metadata: meta,
+				})
+			}
+		}
+	} else {
+		var metricFamily, typ, unit, help string
+		row := conn.QueryRow(context.Background(), "SELECT metric_family, type, unit, help from "+schema.Catalog+".metadata WHERE metric_family=$1 order by last_seen desc limit 1", metric)
+		if err := row.Scan(&metricFamily, &typ, &unit, &help); err != nil {
+			return nil, fmt.Errorf("query result: %w", err)
+		}
+		meta := model.Metadata{
+			// We do not have to send metric name, as its already asked in the params. This
+			// is to keep the behaviour consistent with Prometheus.
+			Unit: unit,
+			Type: typ,
+			Help: help,
+		}
+		seriesId := getSeriesIdsForMetric(metric, metrics, seriesIds)
+		if len(seriesIds) == 0 {
+			// No series exists, meaning we cannot send any target info.
+			// Hence, just send the metadata about the metric.
+			metadataSlice = append(metadataSlice, TargetMetadataType{
+				Metadata: meta,
+			})
+			return metadataSlice, nil
+		}
+		targetInfo, err := getTargetInfo(conn, []string{metric}, [][]model.SeriesID{seriesId})
+		if err != nil {
+			return nil, fmt.Errorf("get target info: %w", err)
+		}
+		targetInfos := targetInfo[metricFamily]
+		for j := range targetInfos {
+			if targetLimit != 0 && len(metadataSlice) >= targetLimit {
+				break
+			}
+			metadataSlice = append(metadataSlice, TargetMetadataType{
+				Target:   j,
+				Metadata: meta,
+			})
+		}
+	}
+	return metadataSlice, nil
+}
+
+func getSeriesIdsForMetric(metric string, metrics []string, seriesIds [][]model.SeriesID) []model.SeriesID {
+	for i := range metrics {
+		if metrics[i] == metric {
+			return seriesIds[i]
+		}
+	}
+	// Corresponding series id does not exists.
+	return []model.SeriesID{}
+}
+
+func getTargetInfo(conn pgxconn.PgxConn, metrics []string, seriesIds [][]model.SeriesID) (map[string]map[Target]struct{}, error) {
```

Comment:
I absoluteley do not understand what this does. Can we explain in a comment?

Reply:
Added comments here and in the line that calls this function.